File: backend/schemas/discovery.py
"""Pydantic schemas for service discovery and batch import."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field

from backend.schemas.enums import ProbeErrorType


# ---------- Probe ----------


class ProbeRequest(BaseModel):
    base_url: str = Field(..., min_length=1, max_length=2048)
    probe_paths: list[str] | None = None


class ProbeAuthConfig(BaseModel):
    base_url: str = Field(..., min_length=1, max_length=2048)
    login_endpoint: str
    login_method: str = "POST"
    login_body: dict[str, Any]
    token_json_path: str | None = None
    probe_paths: list[str] | None = None


class ProbeAttemptSchema(BaseModel):
    path: str
    status: int | None = None
    success: bool = False
    error: str | None = None


class ProbeResultSchema(BaseModel):
    base_url: str
    found: bool = False
    spec_url: str | None = None
    needs_auth: bool = False
    error: str | None = None
    error_type: ProbeErrorType | None = None
    attempts: list[ProbeAttemptSchema] = []


# ---------- NodeDraft ----------


class NodeDraftSchema(BaseModel):
    suggested_name: str
    display_name: str
    description: str = ""
    endpoint: str
    method: str
    input_schema: dict[str, Any] = {}
    output_schema: dict[str, Any] = {}
    category: str | None = None
    tags: list[str] = []
    selected: bool = True


class NodeDraftListResponse(BaseModel):
    base_url: str
    drafts: list[NodeDraftSchema]


# ---------- Batch Import ----------


class BatchImportItem(BaseModel):
    name: str = Field(..., min_length=1, max_length=128)
    display_name: str | None = None
    description: str | None = None
    endpoint: str
    method: str
    input_schema: dict[str, Any] = {}
    output_schema: dict[str, Any] = {}
    category_id: uuid.UUID | None = None
    tags: list[str] = []
    source_path: str | None = None


class BatchImportRequest(BaseModel):
    department_id: uuid.UUID
    credential_id: uuid.UUID | None = None
    base_url: str  # used to build full endpoint = base_url + item.endpoint
    items: list[BatchImportItem] = Field(..., min_length=1)
    # Public fields applied to all imported Nodes
    visibility: str = "internal"
    session_id: uuid.UUID | None = None


class BatchImportResultItem(BaseModel):
    name: str
    node_id: uuid.UUID


class BatchImportResponse(BaseModel):
    imported: int
    nodes: list[BatchImportResultItem]


# ---------- Discovery Session ----------


class DiscoverySessionCreate(BaseModel):
    base_url: str = ""
    source: str = "probe"  # "probe" | "upload"


class DiscoverySessionUpdate(BaseModel):
    status: str | None = None
    spec_url: str | None = None
    total_operations: int | None = None
    imported_count: int | None = None
    completed_at: datetime | None = None


class LinkedNodeSchema(BaseModel):
    id: uuid.UUID
    name: str
    display_name: str | None
    source_path: str | None
    status: str

    class Config:
        from_attributes = True


class DiscoverySessionSchema(BaseModel):
    id: uuid.UUID
    base_url: str
    source: str
    status: str
    spec_url: str | None
    total_operations: int | None
    imported_count: int
    created_at: datetime
    completed_at: datetime | None

    class Config:
        from_attributes = True


class DiscoverySessionDetail(DiscoverySessionSchema):
    nodes: list[LinkedNodeSchema] = []


# ---------- Iteration / Compare ----------


class CompareRequest(BaseModel):
    previous_session_id: uuid.UUID


class CompareResultItem(BaseModel):
    path: str
    method: str
    status: str  # "new" | "imported" | "updated" | "removed"
    changes: dict[str, str] | None = None


class CompareResponse(BaseModel):
    items: list[CompareResultItem]


class IterateActionItem(BaseModel):
    path: str
    method: str
    action: str  # "import" | "update" | "skip"


class IterateRequest(BaseModel):
    actions: list[IterateActionItem] = Field(..., min_length=1)


class IterateResponse(BaseModel):
    imported: int
    updated: int
    skipped: int


class DuplicateUrlResponse(BaseModel):
    duplicate: bool
    existing_session_id: uuid.UUID | None = None
    existing_count: int = 0
    message: str = ""

File: frontend/src/api/discovery.ts
import http from './http'

// ---------- Types ----------

export interface ProbeRequest {
  base_url: string
  probe_paths?: string[]
}

export interface ProbeAuthConfig {
  base_url: string
  login_endpoint: string
  login_method?: string
  login_body: Record<string, unknown>
  token_json_path?: string
  probe_paths?: string[]
}

export interface ProbeAttempt {
  path: string
  status: number | null
  success: boolean
  error: string | null
}

export interface ProbeResult {
  base_url: string
  found: boolean
  spec_url: string | null
  needs_auth: boolean
  error: string | null
  error_type: string | null
  attempts: ProbeAttempt[]
}

export interface NodeDraft {
  suggested_name: string
  display_name: string
  description: string
  endpoint: string
  method: string
  input_schema: Record<string, unknown>
  output_schema: Record<string, unknown>
  category: string | null
  tags: string[]
  selected: boolean
}

export interface NodeDraftListResponse {
  base_url: string
  drafts: NodeDraft[]
}

export interface BatchImportItem {
  name: string
  display_name?: string
  description?: string
  endpoint: string
  method: string
  input_schema?: Record<string, unknown>
  output_schema?: Record<string, unknown>
  category_id?: string
  tags?: string[]
  source_path?: string
}

export interface BatchImportRequest {
  department_id: string
  credential_id?: string
  base_url: string
  items: BatchImportItem[]
  visibility?: string
  session_id?: string
}

export interface BatchImportResultItem {
  name: string
  node_id: string
}

export interface BatchImportResponse {
  imported: number
  nodes: BatchImportResultItem[]
}

// ---------- API ----------

/** Probe a URL for OpenAPI spec */
export const probeSpec = (payload: ProbeRequest) =>
  http.post<ProbeResult | NodeDraftListResponse>('/discovery/probe', payload)

/** Probe with authentication */
export const probeWithAuth = (payload: ProbeAuthConfig) =>
  http.post<ProbeResult | NodeDraftListResponse>('/discovery/probe-with-auth', payload)

/** Upload an OpenAPI spec file */
export const uploadSpec = (file: File) => {
  const formData = new FormData()
  formData.append('file', file)
  return http.post<NodeDraftListResponse>('/discovery/upload-spec', formData, {
    headers: { 'Content-Type': 'multipart/form-data' },
  })
}

/** Batch import nodes from discovered drafts */
export const batchImport = (payload: BatchImportRequest) =>
  http.post<BatchImportResponse>('/discovery/import', payload)

export interface ImportedPathsResponse {
  credential_id: string
  imported_paths: string[]
}

/** Get all source_paths already imported from a given service credential */
export const getImportedPaths = (credentialId: string) =>
  http.get<ImportedPathsResponse>(`/discovery/imported?credential_id=${credentialId}`)


// ---------- Discovery Session ----------

export interface DiscoverySessionCreate {
  base_url?: string
  source?: 'probe' | 'upload'
}

export interface DiscoverySessionUpdate {
  status?: string
  spec_url?: string
  total_operations?: number
  imported_count?: number
  completed_at?: string
}

export interface DiscoverySession {
  id: string
  base_url: string
  source: 'probe' | 'upload'
  status: 'probing' | 'found' | 'failed' | 'completed'
  spec_url: string | null
  total_operations: number | null
  imported_count: number
  created_at: string
  completed_at: string | null
}

export interface LinkedNode {
  id: string
  name: string
  display_name: string | null
  source_path: string | null
  status: string
}

export interface DiscoverySessionDetail extends DiscoverySession {
  nodes: LinkedNode[]
}

/** Create a new discovery session */
export const createSession = (payload: DiscoverySessionCreate = {}) =>
  http.post<DiscoverySession>('/discovery/sessions', payload)

/** Update session status / metadata */
export const updateSession = (sessionId: string, payload: DiscoverySessionUpdate) =>
  http.patch<DiscoverySession>(`/discovery/sessions/${sessionId}`, payload)

/** List discovery sessions for current user */
export const listSessions = (page = 1, pageSize = 20, baseUrl?: string) => {
  const params: Record<string, unknown> = { page, page_size: pageSize }
  if (baseUrl) params.base_url = baseUrl
  return http.get<DiscoverySession[]>('/discovery/sessions', { params })
}

/** Get session detail with linked nodes */
export const getSession = (sessionId: string) =>
  http.get<DiscoverySessionDetail>(`/discovery/sessions/${sessionId}`)

// ---------- Duplicate Detection & Iteration ----------

export interface DuplicateUrlResponse {
  is_duplicate: boolean
  existing_sessions: { id: string; base_url: string; status: string; created_at: string }[]
  existing_node_count: number
}

/** Check if a base_url has already been discovered/imported */
export const checkDuplicate = (baseUrl: string) =>
  http.get<DuplicateUrlResponse>('/discovery/check-duplicate', { params: { base_url: baseUrl } })

export interface CompareResultItem {
  source_path: string
  method: string
  status: 'new' | 'imported' | 'updated' | 'removed'
  node_id: string | null
  display_name: string | null
}

export interface CompareResponse {
  items: CompareResultItem[]
}

/** Compare current session endpoints with previous ones */
export const compareSession = (sessionId: string, previousSessionId: string) =>
  http.post<CompareResponse>(`/discovery/sessions/${sessionId}/compare`, { previous_session_id: previousSessionId })

export interface IterateAction {
  source_path: string
  method: string
  action: 'import' | 'update' | 'skip'
}

export interface IterateResponse {
  imported: number
  updated: number
  skipped: number
}

/** Execute iteration actions on compared endpoints */
export const iterateSession = (sessionId: string, actions: IterateAction[], departmentId: string) =>
  http.post<IterateResponse>(`/discovery/sessions/${sessionId}/iterate`, { actions, department_id: departmentId })
